### PR TITLE
ci(label): skip duplicated build on PR creation

### LINKED ISSUE #3142: Provide more info in `Label/dispatch`
### Problem to Solve

https://github.com/gear-tech/gear/actions/runs/5936484536/job/16096987230?pr=3139

for example, while in-progress and the dispatched checks are just queued, no status displayed in the new created checks

### Possible Solution

log the URL of the new created workflow in `Label/dispatch`

### Notes

_No response_

### LINKED ISSUE #3148: Skip processing build if PR created with `A0-pleaseview`
### Problem to Solve

currently the script is missing checks about if the PR is created with `A0-pleasereview`

### Possible Solution

https://github.com/gear-tech/gear/blob/master/.github/actions/label/build.js#L29

```js
if (check_runs.length === 0) return true
```

### Notes

_No response_

### PR DESCRIPTION
**Pull request by clearloop**
Resolves #3148 
Resolves #3142 

- [x] ( after label checks ) process build only if the previous build is skipped
- [x] mark the name of the triggered builds
- [x] more details of the shadow build
- [x] workaorund for #3154 
- [x] introduce timeout for checks

## Tests

show case of skipping duplicated build: https://github.com/clearloop/gear/pull/2
show case of named builds: https://github.com/gear-tech/gear/actions/runs/5943456240/job/16118640453

@gear-tech/dev

### CODE REVIEW

File: .github/actions/label/build.js [hunk 1]
```diff
@@ -31,12 +31,17 @@ const skip = async ({ github }) => {
...
-  if (runs.length === 0) return false;
+  // Skip this action by default.
+  let skipped = false;
   for (run of runs) {
-    if (run.conclusion !== "skipped") return true;
+    // Process this action only if the previous build has been skipped.
+    if (run.name === "build" && run.conclusion === "skipped") skipped = true;
+
+    // If there is already a build, skip this action without more conditions.
+    if (run.name === "build / linux" || run.name === "build / macox-x86") return true;
```
Comment (shamilsan) [inline .github/actions/label/build.js:41]: Typo?
```suggestion
    if (run.name === "build / linux" || run.name === "build / macos-x86") return true;
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -10,6 +10,15 @@ on:
         type: boolean
         default: false
   workflow_dispatch:
+    inputs:
+      title:
+        type: string
+        description: "Pull request title that triggers this workflow."
+      number:
+        type: string
+        description: "Pull reqeust number that triggers this workflow"
```
Comment (shamilsan) [inline .github/workflows/build.yml:19]:
```suggestion
        description: "Pull request number that triggers this workflow"
```

**Review by shamilsan — APPROVED** [review]

**Comment by shamilsan** [comment]
Need to add [`A4-insubstantial`](https://github.com/gear-tech/gear/labels/A4-insubstantial) to CI triggers too.

File: .github/workflows/label.yml [hunk 3]
```diff
@@ -13,8 +13,9 @@ jobs:
   dispatch:
     runs-on: ubuntu-latest
     if: >-
-      github.event.label.name == 'A0-pleasereview'
-      || github.event.label.name == 'E2-forcemacos'
+      github.event.label.name === 'A0-pleasereview'
```
Comment (breathx) [inline .github/workflows/label.yml:16]: and mergencegreen please

**Review by breathx — APPROVED** [review]

**Review by breathx — APPROVED** [review]

File: .github/workflows/label.yml [hunk 4]
```diff
@@ -13,8 +13,10 @@ jobs:
...
+      || github.event.label.name === 'E2-forcemacos'
+      || github.event.label.name === 'A4-insubstantial'
+      || github.event.label.name === 'A2-mergeongreen'
```
[the remark this answers is not in the document]
Comment (clearloop) [reply .github/workflows/label.yml]: xd

File: .github/actions/label/build.js [hunk 5]
```diff
@@ -100,12 +100,16 @@ const dispatchWorkflow = async ({ core, github }) => {
...
     core.setFailed(`Incorrect workflow runs`);
     return;
   }
 
-  return workflow_runs[0];
+  let sorted_runs = workflow_runs.sort((a, b) => {
+    return new Date(b.created_at) - new Date(a.created_at);
+  })
+
+  return sorted_runs[0]
```
Comment (clearloop) [inline .github/actions/label/build.js:112]: The workflow in #3154 actually has been triggered:

https://github.com/gear-tech/gear/runs/16151228039

But for some unknown reasons, could be `close` + `open` + `label`, that github API could not find the proper place to queue the new created checks, so it failed to be displayed in the checks of the PR, the lines above provide a workaround for it, that if meet that kind of situation:

1. re-run the workflow to clean the previous status
2. re-label can trigger dispatch again

File: .github/actions/label/build.js [hunk 6]
```diff
@@ -63,6 +73,27 @@ const createChecks = async ({ core, github }) => {
...
+      await github.rest.checks.update({
+        owner,
+        repo,
+        check_run_id: check.id,
+        status: "completed",
+        conclusion: "timed_out",
+      });
+    }
+  }, TWO_HOURS)
+}
```
Comment (clearloop) [inline .github/actions/label/build.js:95]: cancel checks when timed out

**Review by ukint-vs — APPROVED** [review]

**Comment by clearloop** [comment]
@CodiumAI-Agent /review

**Comment by clearloop** [comment]
now 12:11, waiting for 2 hrs see if the timeout logic still has bug xd

**Comment by clearloop** [comment]
[f687051](https://github.com/gear-tech/gear/pull/3155/commits/f687051d974c914d14e41a86ef1945f3811d3838) requires rerun implementation #3088

**Pull request merged**